File: src/main/logging/logService.ts
/**
 * 主进程日志服务
 * - 收集主进程所有日志（console 拦截 + 未捕获异常）
 * - 通过 IPC 实时推送到渲染进程
 * - 本地内存环形缓冲区保留最近 5000 条
 */
import { BrowserWindow } from 'electron'
import { randomUUID } from 'crypto'
import type { LogLevel, LogSource, LogEntry } from '../../shared/types'

const MAX_LOG_ENTRIES = 5000

/** 日志服务单例 */
class LogService {
  private buffer: LogEntry[] = []
  private win: BrowserWindow | null = null

  /** 设置主窗口引用，用于推送日志 */
  setWindow(win: BrowserWindow): void {
    this.win = win
  }

  /** 记录一条日志 */
  log(level: LogLevel, source: LogSource, message: string, data?: unknown, stack?: string, origin?: string): void {
    const entry: LogEntry = {
      id: randomUUID(),
      level,
      timestamp: Date.now(),
      source,
      message,
      data: data !== undefined ? (typeof data === 'string' ? data : JSON.stringify(data, null, 2)) : undefined,
      stack,
      origin
    }

    /* 写入缓冲区 */
    this.buffer.push(entry)
    if (this.buffer.length > MAX_LOG_ENTRIES) {
      this.buffer = this.buffer.slice(-MAX_LOG_ENTRIES)
    }

    /* 实时推送到渲染进程 */
    this.pushToRenderer(entry)
  }

  /** 获取所有缓冲日志 */
  getLogs(): LogEntry[] {
    return [...this.buffer]
  }

  /** 清空日志 */
  clear(): void {
    this.buffer = []
  }

  /** 推送单条日志到渲染进程 */
  private pushToRenderer(entry: LogEntry): void {
    try {
      if (this.win && !this.win.isDestroyed()) {
        this.win.webContents.send('log:entry', entry)
      }
    } catch {
      /* 忽略推送错误 */
    }
  }
}

export const logService = new LogService()

File: src-tauri/src/logging/log_service.rs
use std::collections::VecDeque;

use serde::Serialize;
use tauri::{AppHandle, Emitter};
use uuid::Uuid;

/// 单条日志记录
#[derive(Debug, Clone, Serialize, serde::Deserialize)]
pub struct LogEntry {
    pub id: String,
    pub level: String,
    pub timestamp: i64,
    pub source: String,
    pub message: String,
    pub data: Option<String>,
    pub stack: Option<String>,
    pub origin: Option<String>,
}

const MAX_LOG_ENTRIES: usize = 5000;

/// 主进程日志服务 — 环形缓冲 + 实时推送到前端
pub struct LogService {
    buffer: VecDeque<LogEntry>,
    app_handle: Option<AppHandle>,
}

impl Default for LogService {
    fn default() -> Self {
        Self {
            buffer: VecDeque::with_capacity(MAX_LOG_ENTRIES),
            app_handle: None,
        }
    }
}

impl LogService {
    pub fn set_app_handle(&mut self, handle: AppHandle) {
        self.app_handle = Some(handle);
    }

    /// 记录一条日志
    pub fn log(&mut self, level: &str, source: &str, message: &str, data: Option<&str>, stack: Option<&str>, origin: Option<&str>) {
        let entry = LogEntry {
            id: Uuid::new_v4().to_string(),
            level: level.to_string(),
            timestamp: chrono::Utc::now().timestamp_millis(),
            source: source.to_string(),
            message: message.to_string(),
            data: data.map(|s| s.to_string()),
            stack: stack.map(|s| s.to_string()),
            origin: origin.map(|s| s.to_string()),
        };

        // 写入环形缓冲
        if self.buffer.len() >= MAX_LOG_ENTRIES {
            self.buffer.pop_front();
        }
        self.buffer.push_back(entry.clone());

        // 实时推送到前端
        if let Some(ref handle) = self.app_handle {
            let _ = handle.emit("log:entry", &entry);
        }
    }

    pub fn get_logs(&self) -> Vec<LogEntry> {
        self.buffer.iter().cloned().collect()
    }

    pub fn clear(&mut self) {
        self.buffer.clear();
    }
}
